### PR TITLE
Miscellaneous fixes

### LINKED ISSUE #2744: GC\Features\HeapExpansion\Finalizer\Finalizer.cmd Hangs sometimes
Sometimes this test fails in our test runs (it just hangs and XUnit kills it after 10 mins)  I broke into the process once and you could see a deadlock due to trying to finalize during shutdown.

```
  0  Id: 1544.1548 Suspend: 1 Teb: 00007ff7`eaa0e000 Unfrozen
# Child-SP          RetAddr           Call Site
00 0000008e`e299c7a8 00007ffe`fd0213ed ntdll!ZwWaitForMultipleObjects+0xa
01 0000008e`e299c7b0 00007ffe`e7b777b8 KERNELBASE!WaitForMultipleObjectsEx+0xed
02 0000008e`e299ca90 00007ffe`e7b58db7 CoreCLR!WaitForMultipleObjectsEx_SO_TOLERANT(unsigned long nCount = 1, void ** lpHandles = 0x0000008e`e2a06050, int bWaitAll = 0n0, unsigned long dwMilliseconds = 0xffffffff, int bAlertable = 0n1)+0x48 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threads.cpp @ 4279]
03 0000008e`e299cae0 00007ffe`e7b5942d CoreCLR!Thread::DoAppropriateAptStateWait(int numWaiters = 0n1, void ** pHandles = 0x0000008e`e2a06050, int bWaitAll = 0n0, unsigned long timeout = 0xffffffff, WaitMode mode = WaitMode_Alertable (0n1))+0x227 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threads.cpp @ 4313]
04 0000008e`e299cbf0 00007ffe`e7b6f6ff CoreCLR!Thread::DoAppropriateWaitWorker(int countHandles = 0n1, void ** handles = 0x0000008e`e2a06050, int waitAll = 0n0, unsigned long millis = 0xffffffff, WaitMode mode = WaitMode_Alertable (0n1))+0x28d [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threads.cpp @ 4453]
05 0000008e`e299cdd0 00007ffe`e7b58efe CoreCLR!`Thread::DoAppropriateWait'::`11'::__Body::Run(struct Thread::DoAppropriateWait::__l10::__EEParam * __pEEParam = 0x0000008e`e299ce58)+0x8f [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threads.cpp @ 4120]
06 0000008e`e299ce30 00007ffe`e79b3d0e CoreCLR!Thread::DoAppropriateWait(int countHandles = 0n1, void ** handles = 0x0000008e`e2a06050, int waitAll = 0n0, unsigned long millis = 0xffffffff, WaitMode mode = WaitMode_Alertable (0n1), struct PendingSync * syncState = 0x00000000`00000000)+0x10e [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threads.cpp @ 4122]
07 0000008e`e299cec0 00007ffe`e79b31d0 CoreCLR!CLREventBase::WaitEx(unsigned long dwMilliseconds = 0xffffffff, WaitMode mode = WaitMode_Alertable (0n1), struct PendingSync * syncState = 0x00000000`00000000)+0x3ae [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\synch.cpp @ 751]
08 0000008e`e299d030 00007ffe`e7dcd722 CoreCLR!CLREventBase::Wait(unsigned long dwMilliseconds = 0xffffffff, int alertable = 0n1, struct PendingSync * syncState = 0x00000000`00000000)+0x50 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\synch.cpp @ 674]
09 0000008e`e299d070 00007ffe`e7dccfbc CoreCLR!`FinalizerThread::FinalizerThreadWatchDogHelper'::`32'::__Body::Run(struct FinalizerThread::FinalizerThreadWatchDogHelper::__l31::Param * pParam = 0x0000008e`e299d168)+0x62 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\finalizerthread.cpp @ 1381]
0a 0000008e`e299d0b0 00007ffe`e7dccbec CoreCLR!FinalizerThread::FinalizerThreadWatchDogHelper(void)+0x21c [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\finalizerthread.cpp @ 1383]
0b 0000008e`e299d190 00007ffe`e794eae3 CoreCLR!FinalizerThread::FinalizerThreadWatchDog(void)+0x27c [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\finalizerthread.cpp @ 1237]
0c 0000008e`e299d1f0 00007ffe`e794e663 CoreCLR!EEShutDownHelper(int fIsDllUnloading = 0n0)+0x423 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\ceemain.cpp @ 1907]
0d 0000008e`e299dd70 00007ffe`e7974550 CoreCLR!EEShutDown(int fIsDllUnloading = 0n0)+0x3b3 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\ceemain.cpp @ 2461]
0e 0000008e`e299e030 00007ff7`eace3b10 CoreCLR!CorHost2::UnloadAppDomain(unsigned long dwDomainId = 1, int fWaitUntilDone = 0n1)+0x100 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\corhost.cpp @ 3110]
0f 0000008e`e299e070 00007ff7`eace4173 CoreRun!TryRun(int argc = 0n1, wchar_t ** argv = 0x0000008e`e29e9ca8, class Logger * log = 0x0000008e`e299fab4, bool verbose = false, bool waitForDebugger = false, unsigned long * exitCode = 0x0000008e`e299fac8)+0xa60 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\coreclr\hosts\corerun\corerun.cpp @ 633]
10 0000008e`e299fa80 00007ff7`eace59e0 CoreRun!wmain(int argc = 0n2, wchar_t ** argv = 0x0000008e`e29e9ca0)+0x143 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\coreclr\hosts\corerun\corerun.cpp @ 731]
11 (Inline Function) --------`-------- CoreRun!invoke_main+0x22 [f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl @ 89]
12 0000008e`e299fb10 00007ffe`fd3713d2 CoreRun!__scrt_common_main_seh(void)+0x124 [f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl @ 264]
13 0000008e`e299fb50 00007ffe`ffcf5454 KERNEL32!BaseThreadInitThunk+0x22
14 0000008e`e299fb80 00000000`00000000 ntdll!RtlUserThreadStart+0x34

   1  Id: 1544.1064 Suspend: 1 Teb: 00007ff7`eaa0c000 Unfrozen
# Child-SP          RetAddr           Call Site
00 0000008e`e49eefa8 00007ffe`fd0213ed ntdll!ZwWaitForMultipleObjects+0xa
01 0000008e`e49eefb0 00007ffe`e88a3a49 KERNELBASE!WaitForMultipleObjectsEx+0xed
02 0000008e`e49ef290 00007ffe`e88a65aa CoreCLR!DebuggerRCThread::MainLoop(void)+0x739 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\debug\ee\rcthread.cpp @ 1240]
03 0000008e`e49ef540 00007ffe`e88a674f CoreCLR!DebuggerRCThread::ThreadProc(void)+0x7ca [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\debug\ee\rcthread.cpp @ 1043]
04 0000008e`e49ef780 00007ffe`fd3713d2 CoreCLR!DebuggerRCThread::ThreadProcStatic(void * __formal = 0x00000000`00000000)+0xff [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\debug\ee\rcthread.cpp @ 1643]
05 0000008e`e49ef7c0 00007ffe`ffcf5454 KERNEL32!BaseThreadInitThunk+0x22
06 0000008e`e49ef7f0 00000000`00000000 ntdll!RtlUserThreadStart+0x34

   2  Id: 1544.1244 Suspend: 1 Teb: 00007ff7`eaa0a000 Unfrozen
# Child-SP          RetAddr           Call Site
00 0000008e`fd03dbf8 00007ffe`fd021118 ntdll!ZwWaitForSingleObject+0xa
01 0000008e`fd03dc00 00007ffe`e79b0849 KERNELBASE!WaitForSingleObjectEx+0x98
02 0000008e`fd03dca0 00007ffe`e79b2b50 CoreCLR!CLREventWaitHelper2(void * handle = 0x00000000`0000015c, unsigned long dwMilliseconds = 0xffffffff, int alertable = 0n0)+0x49 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\synch.cpp @ 632]
03 0000008e`fd03dce0 00007ffe`e79b08d8 CoreCLR!`CLREventWaitHelper'::`4'::__Body::Run(struct CLREventWaitHelper::__l2::Param * pParam = 0x0000008e`fd03dd58)+0x60 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\synch.cpp @ 658]
04 0000008e`fd03dd20 00007ffe`e79b3d84 CoreCLR!CLREventWaitHelper(void * handle = 0x00000000`0000015c, unsigned long dwMilliseconds = 0xffffffff, int alertable = 0n0)+0x68 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\synch.cpp @ 660]
05 0000008e`fd03dda0 00007ffe`e79b31d0 CoreCLR!CLREventBase::WaitEx(unsigned long dwMilliseconds = 0xffffffff, WaitMode mode = WaitMode_None (0n0), struct PendingSync * syncState = 0x00000000`00000000)+0x424 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\synch.cpp @ 759]
06 0000008e`fd03df10 00007ffe`e7af8726 CoreCLR!CLREventBase::Wait(unsigned long dwMilliseconds = 0xffffffff, int alertable = 0n0, struct PendingSync * syncState = 0x00000000`00000000)+0x50 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\synch.cpp @ 674]
07 0000008e`fd03df50 00007ffe`e7af7ec1 CoreCLR!WKS::gc_heap::user_thread_wait(class CLREventStatic * event = 0x00007ffe`e8efba38, int no_mode_change = 0n0, int time_out_ms = 0n-1)+0x86 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gcee.cpp @ 702]
08 0000008e`fd03df90 00007ffe`e8130b5f CoreCLR!WKS::gc_heap::background_gc_wait(WKS::alloc_wait_reason awr = awr_loh_alloc_during_bgc (0n11), int time_out_ms = 0n-1)+0x101 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gcee.cpp @ 719]
09 0000008e`fd03dfd0 00007ffe`e80c4886 CoreCLR!WKS::gc_heap::wait_for_background(WKS::alloc_wait_reason awr = awr_loh_alloc_during_bgc (0n11))+0x18f [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gc.cpp @ 11909]
0a 0000008e`fd03e000 00007ffe`e8129894 CoreCLR!WKS::gc_heap::allocate_large(int gen_number = 0n3, unsigned int64 size = 0x100038, struct alloc_context * acontext = 0x0000008e`fd03e260, int align_const = 0n7)+0x246 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gc.cpp @ 12554]
0b 0000008e`fd03e160 00007ffe`e80c5ca7 CoreCLR!WKS::gc_heap::try_allocate_more_space(struct alloc_context * acontext = 0x0000008e`fd03e260, unsigned int64 size = 0x100038, int gen_number = 0n3)+0x384 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gc.cpp @ 12891]
0c 0000008e`fd03e1c0 00007ffe`e80c581d CoreCLR!WKS::gc_heap::allocate_more_space(struct alloc_context * acontext = 0x0000008e`fd03e260, unsigned int64 size = 0x100038, int alloc_generation_number = 0n3)+0x27 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gc.cpp @ 13178]
0d 0000008e`fd03e200 00007ffe`e80b35bd CoreCLR!WKS::gc_heap::allocate_large_object(unsigned int64 jsize = 0x100018, int64 * alloc_bytes = 0x0000008e`e2a451e8)+0x16d [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gc.cpp @ 30465]
0e 0000008e`fd03e2b0 00007ffe`e7e4246d CoreCLR!WKS::GCHeap::Alloc(struct alloc_context * acontext = 0x0000008e`e2a451d0, unsigned int64 size = 0x100018, unsigned int flags = 2)+0x24d [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gc.cpp @ 34525]
0f 0000008e`fd03e3f0 00007ffe`e7e431fb CoreCLR!Alloc(unsigned int64 size = 0x100018, int bFinalize = 0n0, int bContainsPointers = 0n16777216)+0x25d [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\gchelpers.cpp @ 106]
10 0000008e`fd03e540 00007ffe`e7bd956f CoreCLR!AllocateArrayEx(class TypeHandle arrayType = class TypeHandle, int * pArgs = 0x0000008e`fd03ea50, unsigned long dwNumArgs = 1, int bAllocateInLargeHeap = 0n0, int bDontSetAppDomain = 0n0)+0xa4b [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\gchelpers.cpp @ 419]
11 0000008e`fd03e9d0 00007ffe`e72191ae CoreCLR!JIT_NewArr1(struct CORINFO_CLASS_STRUCT_ * arrayTypeHnd_ = 0x00007ffe`e6836532, int64 size = 0n131072)+0x64f [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\jithelpers.cpp @ 3379]
12 0000008e`fd03eed0 00007ffe`e72194e5 mscorlib_ni!System.Collections.Generic.List`1[[System.__Canon, mscorlib]].set_Capacity(<HRESULT 0x80004001>)+0x5e [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\mscorlib\src\System\Collections\Generic\List.cs @ 120]
Unable to read dynamic function table entries
Unable to read dynamic function table entries
13 0000008e`fd03ef20 00007ffe`88eb3005 mscorlib_ni!System.Collections.Generic.List`1[[System.__Canon, mscorlib]].Add(<HRESULT 0x80004001>)+0x25 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\mscorlib\src\System\Collections\Generic\List.cs @ 219]
Unable to read dynamic function table entries
Unable to read dynamic function table entries
Unable to read dynamic function table entries
14 0000008e`fd03ef60 00007ffe`e6834422 GCUtil_HeapExpansion!GCUtil.Alloc2(<HRESULT 0x80004001>)+0xf5 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\tests\src\GC\Features\HeapExpansion\GCUtil.cs @ 64]
Unable to read dynamic function table entries
15 0000008e`fd03ef68 00000000`0000000a mscorlib_ni+0xf4422
16 0000008e`fd03ef70 00000000`00000000 0xa

   3  Id: 1544.1eb4 Suspend: 1 Teb: 00007ff7`eaa08000 Unfrozen
# Child-SP          RetAddr           Call Site
00 0000008e`fd49eb18 00007ffe`fd0213ed ntdll!ZwWaitForMultipleObjects+0xa
01 0000008e`fd49eb20 00007ffe`e7b777b8 KERNELBASE!WaitForMultipleObjectsEx+0xed
02 0000008e`fd49ee00 00007ffe`e7b58db7 CoreCLR!WaitForMultipleObjectsEx_SO_TOLERANT(unsigned long nCount = 1, void ** lpHandles = 0x0000008e`e29f0580, int bWaitAll = 0n0, unsigned long dwMilliseconds = 0x13880, int bAlertable = 0n1)+0x48 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threads.cpp @ 4279]
03 0000008e`fd49ee50 00007ffe`e7b5942d CoreCLR!Thread::DoAppropriateAptStateWait(int numWaiters = 0n1, void ** pHandles = 0x0000008e`e29f0580, int bWaitAll = 0n0, unsigned long timeout = 0x13880, WaitMode mode = WaitMode_Alertable (0n1))+0x227 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threads.cpp @ 4313]
04 0000008e`fd49ef60 00007ffe`e7b6f6ff CoreCLR!Thread::DoAppropriateWaitWorker(int countHandles = 0n1, void ** handles = 0x0000008e`e29f0580, int waitAll = 0n0, unsigned long millis = 0x13880, WaitMode mode = WaitMode_Alertable (0n1))+0x28d [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threads.cpp @ 4453]
05 0000008e`fd49f140 00007ffe`e7b58efe CoreCLR!`Thread::DoAppropriateWait'::`11'::__Body::Run(struct Thread::DoAppropriateWait::__l10::__EEParam * __pEEParam = 0x0000008e`fd49f1c8)+0x8f [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threads.cpp @ 4120]
06 0000008e`fd49f1a0 00007ffe`e79b3d0e CoreCLR!Thread::DoAppropriateWait(int countHandles = 0n1, void ** handles = 0x0000008e`e29f0580, int waitAll = 0n0, unsigned long millis = 0x13880, WaitMode mode = WaitMode_Alertable (0n1), struct PendingSync * syncState = 0x00000000`00000000)+0x10e [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threads.cpp @ 4122]
07 0000008e`fd49f230 00007ffe`e79b31d0 CoreCLR!CLREventBase::WaitEx(unsigned long dwMilliseconds = 0x13880, WaitMode mode = WaitMode_Alertable (0n1), struct PendingSync * syncState = 0x00000000`00000000)+0x3ae [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\synch.cpp @ 751]
08 0000008e`fd49f3a0 00007ffe`e7955863 CoreCLR!CLREventBase::Wait(unsigned long dwMilliseconds = 0x13880, int alertable = 0n1, struct PendingSync * syncState = 0x00000000`00000000)+0x50 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\synch.cpp @ 674]
09 0000008e`fd49f3e0 00007ffe`e795560c CoreCLR!WaitForEndOfShutdown_OneIteration(void)+0x223 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\ceemain.cpp @ 1731]
0a 0000008e`fd49f590 00007ffe`e7b80cd1 CoreCLR!WaitForEndOfShutdown(void)+0x19c [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\ceemain.cpp @ 1769]
0b 0000008e`fd49f6b0 00007ffe`e794e0fd CoreCLR!Thread::RareDisablePreemptiveGC(void)+0x6f1 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threadsuspend.cpp @ 3484]
0c 0000008e`fd49f7f0 00007ffe`e7fe4d53 CoreCLR!Thread::DisablePreemptiveGC(void)+0xdd [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threads.h @ 2271]
0d 0000008e`fd49f830 00007ffe`e80e63e6 CoreCLR!GCToEEInterface::DisablePreemptiveGC(class Thread * pThread = 0x0000008e`e2a97860)+0x13 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\gcenv.ee.cpp @ 749]
0e 0000008e`fd49f860 00007ffe`e80c7f69 CoreCLR!WKS::gc_heap::disable_preemptive(class Thread * current_thread = 0x0000008e`e2a97860, int restore_cooperative = 0n1)+0x26 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gc.cpp @ 1791]
0f 0000008e`fd49f890 00007ffe`e80f0f46 CoreCLR!WKS::gc_heap::background_mark_phase(void)+0x6d9 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gc.cpp @ 25510]
10 0000008e`fd49fa40 00007ffe`e80d1cb9 CoreCLR!WKS::gc_heap::gc1(void)+0x176 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gc.cpp @ 15067]
11 0000008e`fd49fb90 00007ffe`e80d200e CoreCLR!WKS::gc_heap::bgc_thread_function(void)+0x9d9 [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gc.cpp @ 26831]
12 0000008e`fd49fc20 00007ffe`e7b7876b CoreCLR!WKS::gc_heap::bgc_thread_stub(void * arg = 0x00000000`00000000)+0x7e [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\gc\gc.cpp @ 24752]
13 0000008e`fd49fc60 00007ffe`fd3713d2 CoreCLR!Thread::intermediateThreadProc(void * arg = 0x0000008e`e2a92090)+0xbb [d:\j\workspace\dotnet_coreclr\debug_windows_nt_prtest\src\vm\threads.cpp @ 2860]
14 0000008e`fd49fe40 00007ffe`ffcf5454 KERNEL32!BaseThreadInitThunk+0x22
15 0000008e`fd49fe70 00000000`00000000 ntdll!RtlUserThreadStart+0x34

#  4  Id: 1544.1ff0 Suspend: 1 Teb: 00007ff7`eaa05000 Unfrozen
# Child-SP          RetAddr           Call Site
00 0000008e`fd61fd68 00007ffe`ffda0034 ntdll!DbgBreakPoint
01 0000008e`fd61fd70 00007ffe`fd3713d2 ntdll!DbgUiRemoteBreakin+0x34
02 0000008e`fd61fda0 00007ffe`ffcf5454 KERNEL32!BaseThreadInitThunk+0x22
03 0000008e`fd61fdd0 00000000`00000000 ntdll!RtlUserThreadStart+0x34
```

I'm going to disable this test against this bug.  I suspect this could be resolved by dotnet/corefx#5205, but it might be good to fix the test so it finishes the finalizes before shutting down.

### PR DESCRIPTION
**Pull request by kouvel**
Related to #2411
- Disabled the assert for CoreCLR. The assert is verifying the circumstances under which this code path is taken on desktop CLR.

Fixes #2744
- Updated the test to deterministically remove the reference to the finalizable object
- Re-enabled the test

Some fixes for issues found with static analysis.

### CODE REVIEW

**Comment by kouvel** [comment]
@sergiy-k @gkhanna79 @stephentoub

**Comment by kouvel** [comment]
Sorry @stephentoub didn't mean to add you :)

File: src/vm/excep.cpp [hunk 1]
```diff
@@ -5675,18 +5676,28 @@ LONG __stdcall COMUnhandledExceptionFilter(     // EXCEPTION_CONTINUE_SEARCH or
...
+        (ExecutionManager::IsManagedCode((PCODE)pExceptionInfo->ExceptionRecord->ExceptionAddress))) // condition 1
+    {
+#ifdef FEATURE_CORRUPTING_EXCEPTIONS
+        if (!CEHelper::IsProcessCorruptedStateException(pExceptionInfo->ExceptionRecord->ExceptionCode)) // condition 4
+        {
+            GCX_COOP();
+            _ASSERTE(CEHelper::IsProcessCorruptedStateException(pThread->GetThrowable())); // condition 4
+        }
+#else
+        _ASSERTE(false);
```
Comment (gkhanna79) [inline src/vm/excep.cpp]: NIT: This assert/#else is not required.

**Comment by gkhanna79** [comment]
LGTM

File: src/vm/excep.cpp (hunk 1 above)
Comment (kouvel) [inline src/vm/excep.cpp]: I was just trying to preserve the previous behavior for that case by having the assertion fail. I'll go ahead and remove it.

Comment (kouvel) [inline src/vm/excep.cpp]: I also moved the ifdef for FEATURE_CORRUPTING_EXCEPTIONS to surround the region.

File: src/vm/excep.cpp [hunk 2]
```diff
@@ -5665,6 +5665,8 @@ LONG __stdcall COMUnhandledExceptionFilter(     // EXCEPTION_CONTINUE_SEARCH or
         return retVal;
     }
 
+#ifndef FEATURE_CORECLR
+#ifdef FEATURE_CORRUPTING_EXCEPTIONS
```
Comment (gkhanna79) [inline src/vm/excep.cpp]: This is not correct. CSE related asserts are the ones that were under the "if" in your original change. This entire assert is not CSE specific.

**Comment by krytarowski** [comment]
Could you please split that into singular patches. 'misc fixes' isn't verbose at all and it will be more difficult to bisect that code.

**Comment by kouvel** [comment]
Sure, I have split it up into a few commits

File: src/vm/excep.cpp (hunk 2 above)
Comment (kouvel) [inline src/vm/excep.cpp]: Hmm I thought that's what I had originally. With this feature disabled, if all of the first four conditions fail, I had an _ASSERTE(false) to fail.

Comment (kouvel) [inline src/vm/excep.cpp]: I have reverted to the original changes for now, please take a look.

File: src/vm/excep.cpp [hunk 3]
```diff
@@ -5683,18 +5684,28 @@ LONG __stdcall COMUnhandledExceptionFilter(     // EXCEPTION_CONTINUE_SEARCH or
...
+        (ExecutionManager::IsManagedCode((PCODE)pExceptionInfo->ExceptionRecord->ExceptionAddress))) // condition 1
+    {
+#ifdef FEATURE_CORRUPTING_EXCEPTIONS
+        if (!CEHelper::IsProcessCorruptedStateException(pExceptionInfo->ExceptionRecord->ExceptionCode)) // condition 4
+        {
+            GCX_COOP();
+            _ASSERTE(CEHelper::IsProcessCorruptedStateException(pThread->GetThrowable())); // condition 4
+        }
+#else // !FEATURE_CORRUPTING_EXCEPTIONS
+        _ASSERTE(false);
```
Comment (gkhanna79) [inline src/vm/excep.cpp]: NIT: You dont need this assert. Otherwise, LGTM.

Comment (kouvel) [inline src/vm/excep.cpp]: Wouldn't removing this _ASSERTE(false) be equivalent to disabling the original assert when FEATURE_CORRUPTING_EXCEPTIONS is disabled? You had mentioned before that the original assert is also valid when the feature is disabled.

**Comment by kouvel** [comment]
@dotnet-bot test this please

File: src/vm/excep.cpp (hunk 3 above)
Comment (gkhanna79) [inline src/vm/excep.cpp]: What I said is correct. Compare the original change (on the left) with the one you are working on the right. You have translated it to a conditional for CSE when, infact, it is doing other assertions still.

Comment (kouvel) [inline src/vm/excep.cpp]: What is CSE?

Suppose FEATURE_CORRUPTING_EXCEPTIONS is off. On the left (before my changes), the code looks like:
```c#
_ASSERTE(A || B || C);
```

With my changes on the right, the code looks like:
```c#
if(!A && !B && !C)
{
    _ASSERTE(false);
}
```

I guess I'm missing something, as those should be equivalent.

**Comment by kouvel** [comment]
@dotnet-bot test this please

File: src/vm/excep.cpp (hunk 3 above)
Comment (kouvel) [inline src/vm/excep.cpp]: @gkhanna79, any further thoughts on this, or should I go ahead and merge?

Comment (gkhanna79) [inline src/vm/excep.cpp]: Please go ahead and merge.

**Pull request merged**